### PR TITLE
Some clean ups, fixes, config flag

### PR DESCRIPTION
**Pull request by coastalwhite**
This PR cleans up some aspects of the configuration management. It also implements #84. This also fixes the long form `page width` flag, which was previously broken.

### CODE REVIEW

**Comment by trimental** [comment]
This changes look good to me. I'll let @CosmicHorrorDev give the final review since this mainly touches upon the code he has written.

**Comment by CosmicHorrorDev** [comment]
I'll give it a look after I'm off work!

File: src/keybindings/defaults.rs [hunk 1]
```diff
@@ -4,98 +4,100 @@ use winit::event::{ModifiersState, VirtualKeyCode as VirtKey};
...
-    } else {
-        ModifiersState::CTRL
-    };
+impl Default for Keybindings {
+    fn default() -> Self {
+        let ctrl_or_command = if IS_MACOS {
+            ModifiersState::LOGO
+        } else {
+            ModifiersState::CTRL
+        };
```
Comment (CosmicHorrorDev) [inline src/keybindings/defaults.rs:13]: I think I'd prefer something like `pub fn defaults() -> Vec<(Action, KeyCombo)>` in this file still and then

```rust
impl Default for Keybindings {
    fn default() -> Self {
        Self(defaults::defaults())
    }
}
```

in mod.rs just to try and keep impls contained in the same file

I definitely like making `Keybindings` into a proper type though!

File: src/keybindings/mod.rs [hunk 2]
```diff
@@ -4,21 +4,42 @@ mod serialization;
...
+pub struct Keybindings(Vec<(Action, KeyCombo)>);
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
 pub enum Key {
     Resolved(VirtualKeyCode),
     ScanCode(ScanCode),
 }
 
+impl Keybindings {
+    #[inline]
```
Comment (CosmicHorrorDev) [inline src/keybindings/mod.rs:22]:
```suggestion
```

Nit: Slapping `#[inline]` on functions without justifying through  something like checking the asm or benchmarks is generally a code smell. Chances are LLVM will already inline this and even if it doesn't this function isn't called nearly enough to be perf sensitive

File: src/opts/config.rs [hunk 3]
```diff
@@ -83,15 +90,25 @@ pub struct Config {
 }
 
 impl Config {
-    pub fn load() -> anyhow::Result<Self> {
-        let config_dir = dirs::config_dir().context("Failed detecting config dir")?;
+    pub fn load_from_file(path: &Path) -> anyhow::Result<Self> {
+        let config_content = read_to_string(path).context(format!(
+            "Failed to read configuration file at '{path}'",
+            path = path.display()
+        ))?;
```
Comment (CosmicHorrorDev) [inline src/opts/config.rs:97]:
```suggestion
        let config_content = read_to_string(path).context(format!(
            "Failed to read configuration file at '{}'",
            path.display(),
        ))?;
```

I think it's clearer to just use the typical interpolation style here

File: src/opts/mod.rs [hunk 4]
```diff
@@ -76,63 +73,76 @@ impl Opts {
...
-            lines_to_scroll: config_lines_to_scroll,
-            light_theme: config_light_theme,
-            dark_theme: config_dark_theme,
-            font_options: config_font_options,
-            keybindings:
-                config::KeybindingsSection {
-                    base: keybindings_base,
-                    extra: keybindings_extra,
-                },
-        } = config;
```
Comment (CosmicHorrorDev) [inline src/opts/mod.rs:100]: I prefer destructuring all the values here as a way of future proofing that any additions to `Config` will be a hard compiler error here. Sure it's more code, but it also helps guide the programmer on what needs to be changed

It really should be done for `args` too. Not sure why I didn't

File: src/opts/cli.rs [hunk 5]
```diff
@@ -43,32 +47,36 @@ pub fn command(scale_help: String, default_theme: Option<ThemeType>) -> Command
...
 
-    let page_width_arg = Arg::new("page_width")
+    let config_arg = Arg::new("config")
+        .short('c')
+        .long("config")
+        .number_of_values(1)
+        .value_parser(value_parser!(PathBuf))
+        .help("Configuration file to use");
+
+    let page_width_arg = Arg::new("page width")
```
Comment (CosmicHorrorDev) [inline src/opts/cli.rs:77]:
```suggestion
    let page_width_arg = Arg::new("page_width")
```

The arg reads a little awkward in the help text with the space in the center (`-w, --page-width <page width>`)

File: src/opts/cli.rs [hunk 6]
```diff
@@ -78,60 +86,38 @@ pub fn command(scale_help: String, default_theme: Option<ThemeType>) -> Command
...
             Self::print_completions(shell, &mut c);
             std::process::exit(0);
         }
 
         let file_path = matches.get_one("file").cloned().expect("required");
         let theme = matches.get_one("theme").cloned();
         let scale = matches.get_one("scale").cloned();
-        let page_width = matches.get_one("page_width").cloned();
+        let config = matches.get_one("config").cloned();
+        let page_width = matches.get_one("page width").cloned();
```
Comment (CosmicHorrorDev) [inline src/opts/cli.rs:114]:
```suggestion
        let page_width = matches.get_one("page_width").cloned();
```

Other part of the last comment

**Review by CosmicHorrorDev — CHANGES_REQUESTED** [review]
Just a few small nits, but otherwise looks great :+1:

This looks like three changes in a trenchcoat, so I'll address each in turn

---

**Fixing the broken page width flag**

Yikes, good find! I'll add a regression test to make sure that it actually gets tested for parsing in a follow-up PR

---

**Various cleanups**

I really appreciate all the changes. Everything seemed great to me!

---

**Adding a `--config` flag**

I see that this dropped modifying default values for `clap`'s help text with values from the config. Technically we _could_ still manage it by trying to parse args -> reading the config file -> actually parse the args, but a) I don't think that's worth the effort and b) I don't even think the original implementation before this PR was worth the effort. It gets a goodbye and good riddance from me :grin:

---

Overall I love the changes! Thanks for the PR

**Comment by coastalwhite** [comment]
I will have a look at this tonight

**Comment by CosmicHorrorDev** [comment]
I'll take a look when I'm on my lunch break in about half an hour

**Review by CosmicHorrorDev — APPROVED** [review]
Thanks for all the changes, everything looks great!

**Pull request merged**